presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayPositionWithIndexFunction.java

```diff
@@ -0,0 +1,146 @@
+    {
+        return arrayPositionWithIndexInternal(type::getDouble, equalMethodHandle, array, element, instance);
+    }
+
+    @TypeParameter("T")
+    @SqlType(StandardTypes.BIGINT)
+    public static long arrayPositionWithIndex(
+            @TypeParameter("T") Type type,
+            @OperatorDependency(operator = EQUAL, argumentTypes = {"T", "T"}) MethodHandle equalMethodHandle,
+            @SqlType("array(T)") Block array,
+            @SqlType("T") Slice element,
+            @SqlType(StandardTypes.BIGINT) long instance)
+    {
+        return arrayPositionWithIndexInternal(type::getSlice, equalMethodHandle, array, element, instance);
+    }
+
+    @TypeParameter("T")
+    @SqlType(StandardTypes.BIGINT)
+    public static long arrayPositionWithIndex(
+            @TypeParameter("T") Type type,
+            @OperatorDependency(operator = EQUAL, argumentTypes = {"T", "T"}) MethodHandle equalMethodHandle,
+            @SqlType("array(T)") Block array,
+            @SqlType("T") Block element,
+            @SqlType(StandardTypes.BIGINT) long instance)
+    {
+        return arrayPositionWithIndexInternal(type::getObject, equalMethodHandle, array, element, instance);
+    }
+
+    protected static long arrayPositionWithIndexInternal(TypeValueExtractor extractor, MethodHandle equalMethodHandle, Block array, Object element, long instance)
+    {
+        int size = array.getPositionCount();
+        int instancesFound = 0;
+
+        if (instance == 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "array_position cannot take a 0-valued instance argument.");
+        }
+
+        int startIndex = instance > 0 ? 0 : size - 1;
+        int stopIndex = instance > 0 ? size : -1;
+        int stepSize = instance > 0 ? 1 : -1;
+        instance = Math.abs(instance);
+
+        for (int i = startIndex; i != stopIndex; i += stepSize) {
+            if (!array.isNull(i)) {
+                try {
+                    Object arrayValue = extractor.extract(array, i);
+                    Boolean result = (Boolean) equalMethodHandle.invoke(arrayValue, element);
+                    checkNotIndeterminate(result);
+                    if (result) {
+                        instancesFound++;
+                        if (instancesFound == instance) {
+                            return i + 1; // result is 1-based (instead of 0)
+                        }
+                    }
+                }
+                catch (Throwable t) {
+                    throw internalError(t);
+                }
+            }
+        }
+        return 0;
+    }
+
+    protected interface TypeValueExtractor
+    {
+        Object extract(Block array, int index);
+    }
+
+    private static void checkNotIndeterminate(Boolean equalsResult)
+    {
+        if (equalsResult == null) {
+            throw new PrestoException(NOT_SUPPORTED, "array_position does not support arrays with elements that are null or contain null");
+        }
+    }
+}
```

Comment:
Hmm - I think we can just return null. Consider this a short hand for an OR which will return null in that situation.

Reply:
Let's discuss changing function behavior separately from extending function APIs. What the function should do when array contains null is not in scope of this PR.

Reply:
Okay. Kept as-is for now.

Reply:
@rongrong / @kaikalur Are any further changes required on this? 

I have just rebased this diff on the latest master. Build failures look transient maven failures, so force-pushing and re-triggering the build.

Reply:
@rongrong / @kaikalur Gentle reminder. Is any further change required on this?